File: Android/app/src/main/java/org/p9/spear/component/ConnectionService.kt
package org.p9.spear.component

import android.net.VpnService
import android.util.Log
import java.io.FileDescriptor

class ConnectionService(
    private val vpn: VpnService,
    private val fd: FileDescriptor,
    private val endpoint: String) {

    private var gateway: IGateway? = null

    fun initialize(): Boolean {
        return try {
            gateway = Gateway(vpn, endpoint)
            gateway?.start(fd)
            true
        } catch (ex: Exception) {
            Log.e(javaClass.name, "ex: $ex")
            false
        }
    }

    fun close() {
        gateway?.stop()
        gateway = null
    }
}

File: iOS/TunnelProvider/Component/ConnectionService.swift
//
//  ConnectionService.swift
//  TunnelProvider
//
//  Created by dexterdreeeam on 2024/1/8.
//

import Foundation
import NetworkExtension

class ConnectionService {
    
    private var endpoint: String
    private var gateway: IGateway?
    
    init(endpoint: String) {
        self.endpoint = endpoint
        self.gateway = nil
    }
    
    func initialize(flow: NEPacketTunnelFlow) -> Bool {
        
        do {
            self.gateway = Gateway(endpoint: self.endpoint)
            self.gateway?.start(flow: flow)
            return true
        } catch {
            return false
        }
    }
    
    func close() {
        self.gateway?.stop()
        self.gateway = nil
    }
}
